runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,387 @@
+        return table.scanAndFilterByEntry(entryPredicate);
+    }
+
+    /**
+     * Start the actual corfu transaction if started by a read operation,
+     * apply all pending operations after starting the corfu transaction.
+     * @param tableBeingRead - only read transactions set this value
+     */
+    private <K extends Message, V extends Message, M extends Message>
+    void txBegin(Table<K, V, M> tableBeingRead) {
+        if (tableBeingRead != null) { // called from one of the read api
+            if (!operations.isEmpty()) { // We have buffered some writes before..
+                operations.forEach(Runnable::run); // apply them for dirty reads to see
+                operations.clear();
+                txnType = TxnType.writeAndReads;
+            }
+            if (txnType == TxnType.writeAndReads) {
+                // Not the first dirty read, don't change the txnType here
+            } else { // Called from read api where no writes have been seen
+                txnType = TxnType.readOnly;
+            }
+            tablesInTxn.add(tableBeingRead);
+        }
+        if (TransactionalContext.isInTransaction()) {
+            if (tableBeingRead == null  // i.e, called from commit() but corfu txn was started!
+                    && !operations.isEmpty()) { // We have pending writes to apply
+                txnType = TxnType.writeAndReads; // Reads, then writes and now commit()
+            }
+            return;
+        }
+        this.txnStartTime = System.nanoTime();
+        Transaction.TransactionBuilder transactionBuilder = this.objectsView
+                .TXBuild()
+                .type(TransactionType.WRITE_AFTER_WRITE);
+        if (isolationLevel.getTimestamp() != Token.UNINITIALIZED) {
+            transactionBuilder.snapshot(isolationLevel.getTimestamp());
+        }
+        transactionBuilder.build().begin();
+    }
+
+    /**
+     * Explicitly end a read only transaction to clean up resources
+     * Internally used at the end of commit() to commit the corfu transaction
+     * @return the address at which the corfu transaction committed.
+     */
+    public long txEnd() {
+        operations.clear();
+        if (TransactionalContext.isInTransaction()) {
+            long commitAddress = this.objectsView.TXEnd();
+            long timeElapsed = System.nanoTime() - txnStartTime;
+            switch (txnType) {
+                case readOnly:
+                    tablesInTxn.forEach(t -> t.readOnlyTxnTimes.set(timeElapsed));
+                    break;
+                case writeOnly:
+                    tablesInTxn.forEach(t -> t.writeOnlyTxnTimes.set(timeElapsed));
+                    break;
+                case writeAndReads:
+                    tablesInTxn.forEach(t -> t.readWriteTxnTimes.set(timeElapsed));
+                    break;
+            }
+            tablesInTxn.clear();
+            return commitAddress;
+        }
+        return Address.NON_ADDRESS;
```

Comment:
should this be an exception or log an error that txEnd was called outside a transaction context?

Reply:
done. thanks